inc/fnsock.h

```diff
@@ -168,6 +168,60 @@ FnSockSetSockOpt(
     _In_ SIZE_T OptionLength
     );
 
+_IRQL_requires_max_(PASSIVE_LEVEL)
+FNSOCK_STATUS
+FnSockGetSockOpt(
+    _In_ FNSOCK_HANDLE Socket,
+    _In_ ULONG Level,
+    _In_ ULONG OptionName,
+    _Out_writes_bytes_(*OptionLength) VOID* OptionValue,
+    _Inout_ SIZE_T* OptionLength
+    );
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+FNSOCK_STATUS
+FnSockIoctl(
+    _In_ FNSOCK_HANDLE Socket,
+    _In_ ULONG ControlCode,
+    _In_reads_bytes_opt_(InputLength) VOID* Input,
+    _In_ ULONG InputLength,
+    _Out_writes_bytes_opt_(OutputLength) VOID* Output,
+    _In_ ULONG OutputLength,
+    _Out_ ULONG* BytesReturned
+    );
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+FNSOCK_STATUS
+FnSockListen(
+    _In_ FNSOCK_HANDLE Socket,
+    _In_ ULONG Backlog
+    );
+
+_IRQL_requires_max_(PASSIVE_LEVEL)
+FNSOCK_HANDLE
+FnSockAccept(
+    _In_ FNSOCK_HANDLE Socket,
+    _Out_writes_bytes_(*AddressLength) struct sockaddr* Address,
+    _Inout_ INT* AddressLength
+    );
+
+FNSOCK_STATUS
+FnSockConnect(
```

Comment:
I'm not sure where best to put this comment, but I recently made a change in XDP's functional test sockets to set so_linger to zero by default, which causes a quick RST when a socket gets closed, rather than letting a socket try graceful shutdown.

The graceful shutdown path can cause subtle issues when a connection persists beyond closesocket() in test code; of course RSTs can, too, but IMO the RST by default is the lesser evil.

Reply:
I added some test code to exercise this. I've confirmed that the default behavior for WSK sockets is abortive shutdown. So there is currently a difference in the behavior of user and kernel FnSock socket default shutdown. We can address this later if needed, but at least XDP is not blocked on this as it can set so_linger to zero for user mode